Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Shadow/DirectionalLightShadowCalculator.azsli

```diff
@@ -316,54 +329,56 @@ float DirectionalShadowCalculator::GetVisibilityFromLightPcf()
     {
         const float3 shadowCoord = m_shadowCoords[indexOfCascade];
-        bool2 checkedShadowed = IsShadowed(shadowCoord, indexOfCascade);
-
         const float distanceMin = ViewSrg::m_esmsDirectional[indexOfCascade].m_lightDistanceOfCameraViewFrustum;
         const float depthDiff = shadowCoord.z - distanceMin;
 
         [branch]
-        if (checkedShadowed.x && depthDiff >= 0)
+        if (depthDiff >= 0)
         {
             const float distanceWithinCameraView = depthDiff / (1. - distanceMin);
             const float3 coord = float3(shadowCoord.xy, indexOfCascade);
             const float occluder = m_directionalLightExponentialShadowmap.SampleLevel(PassSrg::LinearSampler, coord, 0).r;
-            const float exponent = -EsmExponentialShift * (distanceWithinCameraView - occluder);
-            const float ratio = exp(exponent);
-
-            return saturate(ratio);
+            if (occluder < ESM_CASCADE_MARGIN)
+            {
+                const float exponent = -EsmExponentialShift * (distanceWithinCameraView - occluder);
+                const float ratio = exp(exponent);
+                return saturate(ratio);
+            }
         }
     }
     return 1.;
 }
 
 float DirectionalShadowCalculator::GetVisibilityFromLightEsmPcf()
 {
-    for (int indexOfCascade = m_cascadeIndex; indexOfCascade < m_cascadeCount; ++indexOfCascade)
+    for (int indexOfCascade = m_minCascade; indexOfCascade <= m_maxCascade; ++indexOfCascade)
     {
         const float3 shadowCoord = m_shadowCoords[indexOfCascade];
-        bool2 checkedShadowed = IsShadowed(shadowCoord, indexOfCascade);
-
         const float distanceMin = ViewSrg::m_esmsDirectional[indexOfCascade].m_lightDistanceOfCameraViewFrustum;
         const float depthDiff = shadowCoord.z - distanceMin;
 
         [branch]
-        if (checkedShadowed.x && depthDiff >= 0)
+        if (depthDiff >= 0)
         {
             const float distanceWithinCameraView = depthDiff / (1. - distanceMin);
             const float3 coord = float3(shadowCoord.xy, indexOfCascade);
             const float occluder = m_directionalLightExponentialShadowmap.SampleLevel(PassSrg::LinearSampler, coord, 0).r;
-            const float exponent = -EsmExponentialShift * (distanceWithinCameraView - occluder);
-            float ratio = exp(exponent);
-
-            static const float pcfFallbackThreshold = 1.04;
-            if (ratio > pcfFallbackThreshold)
+            
+            if (occluder < ESM_CASCADE_MARGIN)
             {
-                ratio = GetVisibilityFromLightPcf();
+                const float exponent = -EsmExponentialShift * (distanceWithinCameraView - occluder);
+                float ratio = exp(exponent);
+
+                static const float pcfFallbackThreshold = 1.04;
+                if (ratio > pcfFallbackThreshold)
+                {
+                    ratio = GetVisibilityFromLightPcf();
```

Comment:
We are iterating through all the cascades in that function (i.e GetVisibilityFromLightPcf) too. Should we be passing in the index of the cascade for this use case?